### DISCUSSION TITLE
Classification of objects in an image does horribly wrong with yolov8

### DISCUSSION THREAD

**Discussion by mgabell** [ultralytics/ultralytics#8377] [Q&A]
Hi,

I have a dataset that I have trained a resnet50 model and run on our almost daily produced images. The output is very good.
I take the same dataset and prepare for yolo and train from pretrained model. I have tried both M and L models.

The result is horrible. I get a lot of false positive and it does not seem to matter what I do it just keeps getting worse.
I have had a dataset of a very disproportional distribution between classes 13000 of one class and 2000 of another. I have tried to use weights and changed the ration. I reduced the 13000 images to match the other classes. I have always 10% validation images.
I tried the yolov8l model and that did way worse than yolov8m.

Basically I just want something to work 75-80% accurate when running my inference. I always also test on the same set to see if it performs.
I have tried increasing the epoch from 200, to 240 to 2000. This just beats me. It seems the yolo is good in detecting people and phones but not for attached image example.

![image](https://github.com/ultralytics/ultralytics/assets/78966452/64597fb6-e484-427d-9097-334746b6f605)

![image](https://github.com/ultralytics/ultralytics/assets/78966452/2095cb33-4221-4fe8-8c11-313d8db08268)

Those are the two main distinct differences what we want to classify. It just does not. The settings are apart from epochs and imgz=640 the default settings. Running inference the conf is also set to std 0.25.

**Comment by glenn-jocher** [comment]
@mgabell hi there 👋,

I'm sorry to hear you're experiencing issues with object classification using YOLOv8. Let's see if we can improve the performance together!

Given that you've already tried adjusting the dataset balance and training epochs without success, I'd suggest a few more things to consider:

1. **Data Augmentation**: If not already done, try to augment your dataset to include variations like rotations, flips, and color adjustments. This can help the model generalize better.

2. **Hyperparameters**: Tweaking hyperparameters such as learning rate, batch size, or using a different optimizer might yield better results.

3. **Model Architecture**: If YOLOv8 M and L are not performing well, you might want to experiment with the YOLOv8 N model or even custom architectures if you're comfortable with that.

4. **Loss Function**: Investigate the loss function and see if adjusting the weights for different components (like class loss, objectness loss, etc.) helps.

5. **Inference Settings**: At inference time, try adjusting the confidence threshold (`conf`) higher than the default `0.25` to reduce false positives.

Here's a quick example of how you might adjust the confidence threshold during inference:

```python
from ultralytics import YOLO

# Load your trained model
model = YOLO('path/to/your/best.pt')

# Set a higher confidence threshold for inference
results = model.predict('path/to/your/image.jpg', conf=0.5)
```

Remember, achieving high accuracy can sometimes be a bit of trial and error, especially with a new dataset. If you continue to face difficulties, feel free to share more details, such as loss curves or specific examples of misclassifications, and we can dive deeper into the issue.

Keep experimenting, and don't hesitate to reach out for more assistance! 🚀

**Reply by mgabell** [reply]
Thanks for reaching out. 
Data Augmentation. Is this possible with ultralytics? Or do I have to do it manually?
How can I use a custom architecture? I assume, forgive my ignorance you mean train from scratch?
Thirdly, Why should I drop down to N? Isn't that a smaller model? Is it because I dropped the dataset to 3-4000 images, that should go down in size?
This is my train command:
yolo detect train model=yolov8n.pt data=dataset.yaml epochs=100 batch=32 imgsz=640 optimizer=SGD
I start now training with SGD and change optimizer as I go but leave everything else. Perhaps chaning batch size too

**Reply by glenn-jocher** [reply]
@mgabell hey there! 👋

Absolutely, data augmentation is built into the Ultralytics YOLO training pipeline, so no need for manual preprocessing. The augmentations like rotations, flips, and color adjustments are automatically applied during training.

For custom architectures, it's not necessarily about training from scratch. You can modify an existing `.yaml` architecture file to better suit your dataset. It's like tweaking the model's layers and parameters to find the sweet spot for your specific use case.

Dropping down to YOLOv8 N could be beneficial because smaller models can sometimes generalize better on smaller datasets. It's not a hard rule, but it's worth a try if the larger models are overfitting.

Your train command looks good! 👍 SGD is a solid choice to start with. Adjusting the batch size can also help, especially if you're monitoring your GPU memory usage.

Keep experimenting, and you'll find the right setup. Happy training! 🚀
